diff --git a/org/mvel2/optimizers/impl/asm/ASMAccessorOptimizer.java b/org/mvel2/optimizers/impl/asm/ASMAccessorOptimizer.java
index 1887c0b..e11c45e 100644
--- a/org/mvel2/optimizers/impl/asm/ASMAccessorOptimizer.java
+++ b/org/mvel2/optimizers/impl/asm/ASMAccessorOptimizer.java
@@ -85,7 +85,8 @@ public class ASMAccessorOptimizer extends AbstractOptimizer implements AccessorO
       OPCODES_VERSION = Opcodes.V1_4;
     else if (javaVersion.startsWith("1.5"))
       OPCODES_VERSION = Opcodes.V1_5;
-    else if (javaVersion.startsWith("1.6") || javaVersion.startsWith("1.7"))
+    else if (javaVersion.startsWith("1.6") || javaVersion.startsWith("1.7")
+             || javaVersion.startsWith("1.8"))
       OPCODES_VERSION = Opcodes.V1_6;
     else
       OPCODES_VERSION = Opcodes.V1_2;
@@ -3275,3 +3276,4 @@ public class ASMAccessorOptimizer extends AbstractOptimizer implements AccessorO
     return literal;
   }
 }
+/* @generated */
\ No newline at end of file
