diff --git a/org/mvel2/optimizers/impl/asm/ASMAccessorOptimizer.java b/org/mvel2/optimizers/impl/asm/ASMAccessorOptimizer.java
index e11c45e84245..35e985316ee9 100644
--- a/org/mvel2/optimizers/impl/asm/ASMAccessorOptimizer.java
+++ b/org/mvel2/optimizers/impl/asm/ASMAccessorOptimizer.java
@@ -81,16 +81,13 @@ public class ASMAccessorOptimizer extends AbstractOptimizer implements AccessorO
 
   static {
     final String javaVersion = getProperty("java.version");
-    if (javaVersion.startsWith("1.4"))
+    if (javaVersion.startsWith("1.4")) {
       OPCODES_VERSION = Opcodes.V1_4;
-    else if (javaVersion.startsWith("1.5"))
+    } else if (javaVersion.startsWith("1.5")) {
       OPCODES_VERSION = Opcodes.V1_5;
-    else if (javaVersion.startsWith("1.6") || javaVersion.startsWith("1.7")
-             || javaVersion.startsWith("1.8"))
+    } else {
       OPCODES_VERSION = Opcodes.V1_6;
-    else
-      OPCODES_VERSION = Opcodes.V1_2;
-
+    }
 
     String defaultNameSapce = getProperty("mvel2.namespace");
     if (defaultNameSapce == null) NAMESPACE = "org/mvel2/";
