diff --git a/freemarker/ext/jsp/JspTagModelBase.java b/freemarker/ext/jsp/JspTagModelBase.java
index 032cead0ab0e..bab511a700bf 100644
--- a/freemarker/ext/jsp/JspTagModelBase.java
+++ b/freemarker/ext/jsp/JspTagModelBase.java
@@ -19,10 +19,6 @@
 
 package freemarker.ext.jsp;
 
-import java.beans.BeanInfo;
-import java.beans.IntrospectionException;
-import java.beans.Introspector;
-import java.beans.PropertyDescriptor;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.math.BigDecimal;
@@ -50,16 +46,24 @@ class JspTagModelBase {
     private final Method dynaSetter;
     private final Map propertySetters = new HashMap();
     
-    protected JspTagModelBase(String tagName, Class tagClass) throws IntrospectionException {
+    protected JspTagModelBase(String tagName, Class tagClass) {
         this.tagName = tagName;
         this.tagClass = tagClass;
-        BeanInfo bi = Introspector.getBeanInfo(tagClass);
-        PropertyDescriptor[] pda = bi.getPropertyDescriptors();
-        for (int i = 0; i < pda.length; i++) {
-            PropertyDescriptor pd = pda[i];
-            Method m = pd.getWriteMethod();
-            if (m != null) {
-                propertySetters.put(pd.getName(), m);
+        for (Method method : tagClass.getMethods()) {
+            if (method.getReturnType().equals(void.class)
+                && (method.getParameterTypes().length == 1)) {
+
+                String methodName = method.getName();
+
+                if (methodName.startsWith("set") && (methodName.length() > 3) &&
+                    Character.isUpperCase(methodName.charAt(3))) {
+
+                    String fieldName =
+                        methodName.substring(3, 4).toLowerCase()
+                        + methodName.substring(4);
+
+                    propertySetters.put(fieldName, method);
+                }
             }
         }
         // Check to see if the tag implements the JSP2.0 DynamicAttributes
@@ -167,3 +171,4 @@ class JspTagModelBase {
     }
     
 }
+/* @generated */
\ No newline at end of file
